src/test/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionSpec.scala

```diff
@@ -0,0 +1,153 @@
+      removed.foreach { email => verify(mockGoogleDirectoryDAO).removeMemberFromGroup(target.email, WorkbenchUserEmail(email)) }
+      verify(mockDirectoryDAO).updateSynchronizedDate(target.id)
+    }
+  }
+
+  "GoogleExtension" should "get a pet service account for a user" in {
+    implicit val patienceConfig = PatienceConfig(1 second)
+    val dirDAO = new JndiDirectoryDAO(directoryConfig)
+    val schemaDao = new JndiSchemaDAO(directoryConfig)
+    runAndWait(schemaDao.init())
+    try {
+      val mockGoogleIamDAO = new MockGoogleIamDAO
+      val mockGoogleDirectoryDAO = new MockGoogleDirectoryDAO
+
+      val googleExtensions = new GoogleExtensions(dirDAO, null, mockGoogleDirectoryDAO, null, mockGoogleIamDAO, googleServicesConfig, petServiceAccountConfig)
+      val service = new UserService(dirDAO, googleExtensions, mockGoogleDirectoryDAO, googleServicesConfig.appsDomain)
+
+      val defaultUserId = WorkbenchUserId("newuser")
+      val defaultUserEmail = WorkbenchUserEmail("newuser@new.com")
+      val defaultUser = WorkbenchUser(defaultUserId, defaultUserEmail)
+
+      // create a user
+      val newUser = service.createUser(defaultUser).futureValue
+      newUser shouldBe UserStatus(UserStatusDetails(defaultUserId, defaultUserEmail), Map("ldap" -> true, "allUsersGroup" -> true, "google" -> true))
+
+      // create a pet service account
+      val emailResponse = googleExtensions.createUserPetServiceAccount(defaultUser).futureValue
+
+      emailResponse.value should endWith("@my-pet-project.iam.gserviceaccount.com")
+
+      // verify ldap
+      dirDAO.getPetServiceAccountForUser(defaultUserId).futureValue shouldBe Some(emailResponse)
+
+      val ldapPetOpt = dirDAO.loadSubjectFromEmail(emailResponse.value).flatMap {
+        case Some(subject: WorkbenchUserServiceAccountSubjectId) => dirDAO.loadPetServiceAccount(subject)
+        case _ => fail(s"could not load pet LDAP entry from $emailResponse")
+      }.futureValue
+
+      ldapPetOpt shouldBe 'defined
+      val Some(ldapPet) = ldapPetOpt
+      ldapPet.email shouldBe WorkbenchUserServiceAccountEmail(emailResponse.value)
+      ldapPet.displayName shouldBe WorkbenchUserServiceAccountDisplayName("")
+      // MockGoogleIamDAO generates the subject ID as a random Long
+      Try(ldapPet.subjectId.value.toLong) shouldBe a[Success[_]]
+
+      // verify google
+      val groupEmail = WorkbenchGroupEmail(googleExtensions.toProxyFromUser(defaultUserId.value))
+      mockGoogleIamDAO.serviceAccounts should contain key (emailResponse)
+      mockGoogleDirectoryDAO.groups should contain key (groupEmail)
+      mockGoogleDirectoryDAO.groups(groupEmail) shouldBe Set(defaultUserEmail, emailResponse)
+
+      // create one again, it should work
+      val petSaResponse2 = googleExtensions.createUserPetServiceAccount(defaultUser).futureValue
+      petSaResponse2 shouldBe emailResponse
+    } finally {
+      runAndWait(schemaDao.clearDatabase())
```

Comment:
If it's worth clearing the database at the end, you might want to also clear it at the beginning just in case something went wrong and with the cleanup during a previous test run.

Reply:
would it be better to just have a standard one way or the other. So far we just clean up at the end. Maybe it is safer to start clean.

Reply:
Could do only pre-cleanup... it depends on what state you want to leave the system. My recent experience is with things like the selenium test for registration where I do want to clean up after the test to leave Google in the expected rest state but also clean up before the test to avoid test failures due to previous failed clean up (which might then only be able to be resolved via manual intervention).

Reply:
there is not an external system to worry about here